pkg/controllers/management/rkeworkerupgrader/plan.go

```diff
@@ -0,0 +1,355 @@
+package rkeworkerupgrader
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+	kd "github.com/rancher/rancher/pkg/controllers/management/kontainerdrivermetadata"
+	"github.com/rancher/rancher/pkg/librke"
+	nodeserver "github.com/rancher/rancher/pkg/rkenodeconfigserver"
+	rkehosts "github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/services"
+	rkeservices "github.com/rancher/rke/services"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+func (uh *upgradeHandler) nonWorkerPlan(node *v3.Node, cluster *v3.Cluster) (*v3.RKEConfigNodePlan, error) {
+	rkeConfig := cluster.Status.AppliedSpec.RancherKubernetesEngineConfig
+
+	if rkeConfig == nil {
+		rkeConfig = &v3.RancherKubernetesEngineConfig{}
+	}
+
+	rkeConfig = rkeConfig.DeepCopy()
+	rkeConfig.Nodes = []v3.RKEConfigNode{
+		*node.Status.NodeConfig,
+	}
+	rkeConfig.Nodes[0].Role = []string{rkeservices.WorkerRole, rkeservices.ETCDRole, rkeservices.ControlRole}
+
+	infos, err := librke.GetDockerInfo(node)
+	if err != nil {
+		return nil, err
+	}
+
+	hostAddress := node.Status.NodeConfig.Address
+	hostDockerInfo := infos[hostAddress]
+
+	svcOptions, err := uh.getServiceOptions(rkeConfig.Version, hostDockerInfo.OSType)
+	if err != nil {
+		return nil, err
+	}
+
+	plan, err := librke.New().GeneratePlan(uh.ctx, rkeConfig, infos, svcOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := uh.systemAccountManager.GetOrCreateSystemClusterToken(cluster.Name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create or get cluster token for share-mnt")
+	}
+
+	np := &v3.RKEConfigNodePlan{}
+
+	for _, tempNode := range plan.Nodes {
+		if tempNode.Address == hostAddress {
+
+			b2d := strings.Contains(infos[tempNode.Address].OperatingSystem, rkehosts.B2DOS)
+			np.Processes = nodeserver.AugmentProcesses(token, tempNode.Processes, false, b2d,
+				node.Status.NodeConfig.HostnameOverride, cluster)
+
+			np.Processes = nodeserver.AppendTaintsToKubeletArgs(np.Processes, node.Status.NodeConfig.Taints)
+
+			return np, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find plan for %s", hostAddress)
+}
+
+func (uh *upgradeHandler) workerPlan(node *v3.Node, cluster *v3.Cluster) (*v3.RKEConfigNodePlan, error) {
+	infos, err := librke.GetDockerInfo(node)
+	if err != nil {
+		return nil, err
+	}
+
+	hostAddress := node.Status.NodeConfig.Address
+	hostDockerInfo := infos[hostAddress]
+
+	rkeConfig := cluster.Status.AppliedSpec.RancherKubernetesEngineConfig.DeepCopy()
```

Comment:
need to check for RancherKubernetesEngineConfig being nil here too?

Reply:
think this should cover it all, https://github.com/rancher/rancher/pull/25451/files#diff-a16c43ed026936b620e2fdb9ca6196b8R98